store/src/main/java/org/apache/rocketmq/store/schedule/ScheduleMessageService.java

```diff
@@ -308,158 +387,408 @@ public void executeOnTimeup() {
+            if (this.autoResend) {
+                this.resend();
+            } else {
+                this.status = ProcessStatus.SKIP;
+            }
+        }
+
+        public ProcessStatus getStatus() {
+            return this.status;
+        }
+
+        public PutMessageResult get() {
+            try {
+                return this.future.get();
+            } catch (InterruptedException | ExecutionException e) {
+                return new PutMessageResult(PutMessageStatus.UNKNOWN_ERROR, null);
+            }
+        }
+
+        private void resend() {
+            log.info("Resend message, info: {}", this.toString());
+
+            // Gradually increase the resend interval.
+            try {
+                Thread.sleep(Math.min(this.resendCount++ * 100, 60 * 1000));
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+
+            try {
+                MessageExt msgExt = ScheduleMessageService.this.defaultMessageStore.lookMessageByOffset(this.physicOffset, this.physicSize);
+                if (msgExt == null) {
+                    log.warn("ScheduleMessageService resend not found message. info: {}", this.toString());
+                    this.status = need2Skip() ? ProcessStatus.SKIP : ProcessStatus.EXCEPTION;
+                    return;
+                }
+
+                MessageExtBrokerInner msgInner = ScheduleMessageService.this.messageTimeup(msgExt);
+                PutMessageResult result = ScheduleMessageService.this.writeMessageStore.putMessage(msgInner);
+                this.handleResult(result);
+                if (result != null && result.getPutMessageStatus() == PutMessageStatus.PUT_OK) {
+                    log.info("Resend message success, info: {}", this.toString());
+                }
+            } catch (Exception e) {
+                this.status = ProcessStatus.EXCEPTION;
+                log.error("Resend message error, info: {}", this.toString(), e);
+            }
+        }
+
+        public boolean need2Blocked() {
+            return this.resendCount > ScheduleMessageService.this.maxResendNum2Blocked;
+        }
+
+        public boolean need2Skip() {
+            return this.resendCount > ScheduleMessageService.this.maxResendNum2Blocked * 2;
+        }
+
+        @Override
+        public String toString() {
+            return "PutResultProcess{" +
+                "topic='" + topic + '\'' +
+                ", offset=" + offset +
+                ", physicOffset=" + physicOffset +
+                ", physicSize=" + physicSize +
+                ", delayLevel=" + delayLevel +
+                ", msgId='" + msgId + '\'' +
+                ", autoResend=" + autoResend +
+                ", resendCount=" + resendCount +
+                ", status=" + status +
+                '}';
+        }
+    }
+
+    public enum ProcessStatus {
+        RUNNING,
+        SUCCESS,
+        EXCEPTION,
+        SKIP,
     }
```

Comment:
It would be better to add comments to explain the meaning of each state and the message processing results. For example, If return ProcessStatus.SKIP, it will discard the message?

Reply:
Ok, I will add a comment description. 
In some extreme cases, in order to avoid blocking, message delivery will be skipped. For example: 
After the message delivery fails, a retry will be performed at this time. If the message does not exist (the message expires), the delivery will be skipped to prevent the entire delay level from being blocked and unavailable.